crates/notion-core/src/inventory/mod.rs

```diff
@@ -135,36 +130,61 @@ impl<D: Distro> Collection<D> {
 }
 
 pub trait FetchResolve<D: Distro> {
+    type FetchedVersion;
+
     /// Fetch a Distro version matching the specified semantic versioning requirements.
     fn fetch(
         &mut self,
+        name: String, // unused by Node and Yarn, but package install needs this
         matching: &VersionSpec,
         hooks: Option<&ToolHooks<D>>,
-    ) -> Fallible<Fetched<DistroVersion>>;
+    ) -> Fallible<Fetched<Self::FetchedVersion>>;
 
     /// Resolves the specified semantic versioning requirements into a distribution
-    fn resolve(&self, matching: &VersionSpec, hooks: Option<&ToolHooks<D>>) -> Fallible<D> {
+    fn resolve(
+        &self,
+        name: String,
+        matching: &VersionSpec,
+        hooks: Option<&ToolHooks<D>>,
+    ) -> Fallible<D> {
         let version = match *matching {
-            VersionSpec::Latest => self.resolve_latest(hooks)?,
-            VersionSpec::Semver(ref requirement) => self.resolve_semver(requirement, hooks)?,
-            VersionSpec::Exact(ref version) => version.clone(),
+            VersionSpec::Latest => self.resolve_latest(name.clone(), hooks)?,
+            VersionSpec::Semver(ref requirement) => {
+                self.resolve_semver(name.clone(), requirement, hooks)?
+            }
+            VersionSpec::Exact(ref version) => {
+                self.resolve_exact(name.clone(), version.clone(), hooks)?
+            }
         };
 
-        D::new(version, hooks)
+        D::new(name, version, hooks)
     }
 
     /// Resolves the latest version for this tool, using either the `latest` hook or the public registry
-    fn resolve_latest(&self, hooks: Option<&ToolHooks<D>>) -> Fallible<Version>;
+    fn resolve_latest(
+        &self,
+        name: String,
+        hooks: Option<&ToolHooks<D>>,
+    ) -> Fallible<D::ResolvedVersion>;
 
     /// Resolves a SemVer version for this tool, using either the `index` hook or the public registry
     fn resolve_semver(
         &self,
+        name: String,
         matching: &VersionReq,
         hooks: Option<&ToolHooks<D>>,
-    ) -> Fallible<Version>;
+    ) -> Fallible<D::ResolvedVersion>;
+
+    /// Resolves an exact version of this tool
+    fn resolve_exact(
+        &self,
+        name: String,
+        version: Version,
+        hooks: Option<&ToolHooks<D>>,
+    ) -> Fallible<D::ResolvedVersion>;
 }
 
-fn registry_fetch_error(error: &reqwest::Error) -> ErrorDetails {
+pub fn registry_fetch_error(error: &reqwest::Error) -> ErrorDetails {
```

Comment:
This is another function which is only used internally to the module, does it need to be made `pub`?

Reply:
Good catch, will fix this one too